File: scala/src/main/scala/com/stulsoft/pvertx/scala/JobVerticle.scala
/*
 * Copyright (c) 2018. Yuriy Stul
 */

package com.stulsoft.pvertx.scala

import io.vertx.lang.scala.ScalaVerticle
import io.vertx.scala.core.Vertx
import org.slf4j.LoggerFactory

import scala.concurrent.Future
import scala.util.{Failure, Success}

/**
  * @author Yuriy Stul
  */
class JobVerticle extends ScalaVerticle {
  private val logger = LoggerFactory.getLogger(classOf[JobVerticle])

  override def stopFuture(): Future[_] = {
    logger.info("Stopping")
    Future.successful()
  }

  override def startFuture(): Future[_] = {
    logger.info("Stopping")
    super.startFuture()
//    Future.failed(new RuntimeException("test error"))
  }
}

object JobVerticle extends App {
  private val logger = LoggerFactory.getLogger(classOf[JobVerticle])

  import scala.concurrent.ExecutionContext.Implicits.global

  var vertx = Vertx.vertx
  vertx.deployVerticleFuture(ScalaVerticle.nameForVerticle[JobVerticle])
    .onComplete {
      case Success(result) =>
        logger.info("Success, verticle ID is {}", result)
        vertx.setTimer(1000, l => vertx.close)
      case Failure(t) =>
        logger.error("Failure {}", t.getMessage)
        vertx.close()
    }
}

File: pqueue/src/main/java/com/stulsoft/pqueue/JobVerticle.java
/*
 * Copyright (c) 2020. Yuriy Stul
 */

package com.stulsoft.pqueue;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.eventbus.Message;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Random;

/**
 * @author Yuriy Stul
 */
public class JobVerticle extends AbstractVerticle {
    private static final Logger logger = LoggerFactory.getLogger(JobVerticle.class);

    public static String EB_ADDRESS = "JobVerticle";

    private final Random random = new Random();

//    private int[] delays={500, 700, 800, 1000};
    private int[] delays={100, 200, 500, 700, 800, 1000};

    @Override
    public void start(Future<Void> startFuture) {
        logger.info("Starting JobVerticle ...");
        vertx.eventBus().consumer(EB_ADDRESS, this::handler);
        startFuture.complete();
    }

    @Override
    public void stop(Future<Void> stopFuture) {
        logger.info("Stopping JobVerticle ...");
        stopFuture.complete();
    }

    private void handler(Message<String> msg) {
        logger.info("Handling [{}] message", msg.body());
        vertx.setTimer(defineDelay(), l -> msg.reply("Done with " + msg.body()));
    }

    private long defineDelay() {
        return delays[random.nextInt(delays.length)];
    }
}
